File: src/main/java/org/c3lang/intellij/C3LineMarkerProvider.java
package org.c3lang.intellij;

import com.intellij.codeInsight.daemon.LineMarkerInfo;
import com.intellij.codeInsight.daemon.LineMarkerProvider;
import com.intellij.execution.RunManager;
import com.intellij.execution.configurations.ConfigurationFactory;
import com.intellij.execution.configurations.ConfigurationTypeUtil;
import com.intellij.execution.executors.DefaultRunExecutor;
import com.intellij.execution.runners.ExecutionEnvironment;
import com.intellij.execution.runners.ExecutionEnvironmentBuilder;
import com.intellij.icons.AllIcons;
import com.intellij.openapi.editor.markup.GutterIconRenderer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.execution.ExecutionException;
import org.c3lang.intellij.psi.C3FuncDefinition;
import org.jetbrains.annotations.Nullable;

public class C3LineMarkerProvider implements LineMarkerProvider
{
	@Override
	public @Nullable LineMarkerInfo<?> getLineMarkerInfo(PsiElement element)
	{
		if (!(element instanceof C3FuncDefinition function)) return null;

		String type = function.getFuncDef().getFuncHeader().getOptionalType().getType().getText();
		String name = function.getFuncDef().getFqName().getName();

		if (!name.equals("main")) return null;
		if (!type.equals("int") && !type.equals("void")) return null;

		return new LineMarkerInfo<>(
			element,
			element.getTextRange(),
			AllIcons.Actions.Execute,
			ignored -> "Run main",
			(event, elt) -> createAndRunCustomConfig(elt.getProject()),
			GutterIconRenderer.Alignment.RIGHT,
			() -> "Click to open context menu"
		);
	}

	private static void createAndRunCustomConfig(Project project)
	{
		RunManager runManager = RunManager.getInstance(project);
		C3BuildRunConfigurationType configType =
			ConfigurationTypeUtil.findConfigurationType(C3BuildRunConfigurationType.class);
		ConfigurationFactory factory = configType.getConfigurationFactories()[0];
		var settings = runManager.createConfiguration("main", factory);
		C3BuildRunConfiguration config = (C3BuildRunConfiguration) settings.getConfiguration();

		config.setWorkingDirectory(project.getBasePath());
		config.setArgs("");

		runManager.addConfiguration(settings);
		runManager.setSelectedConfiguration(settings);

		var executor = DefaultRunExecutor.getRunExecutorInstance();
		try
		{
			ExecutionEnvironment environment = ExecutionEnvironmentBuilder.create(executor, settings).build();
			environment.getRunner().execute(environment);
		}
		catch (ExecutionException e)
		{
			throw new RuntimeException(e);
		}
	}
}

File: src/main/java/org/c3lang/intellij/C3LineMarkerProvider.kt
package org.c3lang.intellij

import com.intellij.codeInsight.daemon.LineMarkerInfo
import com.intellij.codeInsight.daemon.LineMarkerProvider
import com.intellij.execution.RunManager
import com.intellij.execution.configurations.ConfigurationTypeUtil
import com.intellij.execution.executors.DefaultRunExecutor
import com.intellij.execution.runners.ExecutionEnvironmentBuilder
import com.intellij.icons.AllIcons
import com.intellij.openapi.editor.markup.GutterIconRenderer
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import org.c3lang.intellij.psi.C3FuncDefinition

class C3LineMarkerProvider : LineMarkerProvider
{
    override fun getLineMarkerInfo(element: PsiElement): LineMarkerInfo<*>?
    {
        if (element !is C3FuncDefinition) return null

        val type = element.funcDef.funcHeader.optionalType.type.text
        val name = element.funcDef.fqName.name

        if (name != "main") return null
        if (type != "int") return null

        return LineMarkerInfo(
            element,
            element.textRange,
            AllIcons.Actions.Execute,
            { "Run main" },
            { _, elt -> createAndRunCustomConfig(elt.project) },
            GutterIconRenderer.Alignment.RIGHT,
            { "Click to open context menu" }
        )
    }

    private fun createAndRunCustomConfig(project: Project)
    {
        val runManager = RunManager.getInstance(project)
        val configType = ConfigurationTypeUtil.findConfigurationType(C3BuildRunConfigurationType::class.java)
        val factory = configType.configurationFactories.first()
        val uniqueName = "main"
        val settings = runManager.createConfiguration(uniqueName, factory)
        val config = settings.configuration as C3BuildRunConfiguration

        config.workingDirectory = project.basePath
        config.args = ""

        runManager.addConfiguration(settings)
        runManager.selectedConfiguration = settings

        val executor = DefaultRunExecutor.getRunExecutorInstance()
        val environment = ExecutionEnvironmentBuilder.create(executor, settings).build()

        environment.runner.execute(environment)
    }
}